docs/advanced/how-it-works.md

```diff
@@ -0,0 +1,175 @@
+      }
+    ]
+  },
+  "signature": {
+    "type": "SHA3MerkleProof",
+    "targetHash": "11d456db211d68cc8a6eac5e293422dec669b54812e4975497d7099467335987",
+    "proof": [],
+    "merkleRoot": "11d456db211d68cc8a6eac5e293422dec669b54812e4975497d7099467335987"
+  }
+}
+```
+
+Multiple interesting things happened that we will dig in:
+
+- A `data` has been created and holds the content of the file provided for wrapping with some weird-looking extra data.
+- A `signature` object has been created.
+
+### The `data` object
+
+The first step of wrapping consists of running through and transforming all the object properties provided as input:
+
+1. For each properties
+1. Generate a salt [using uuid v4](https://www.npmjs.com/package/uuid) in order to prevent [rainbow table attack](https://en.wikipedia.org/wiki/Rainbow_table).
+1. Find the type original property.
+1. Transform the original value to `<salt>:<original-type>:<original-value>`.
+
+> The shape of the input object remains untouched.
+
+### The `signature` object
+
+#### targetHash
+
+Once the `data` object has been computed we will be able to create an unique hash for the document that we will set into `targetHash`:
+
+1. List each properties' path from the `data` object and associate its value. The path follows the [flatley](https://github.com/antony/flatley) path convention.
+1. For each properties' path, compute a hash using the properties' path and value. To compute the hash we use [keccack256](https://en.wikipedia.org/wiki/SHA-3).
+1. Sort all the hashes from the previous step alphabetically and hash them all together: this will provide the `targetHash` of the document. To compute the `targetHash` we also use [keccack256](https://en.wikipedia.org/wiki/SHA-3).
+
+> The `targetHash` of a document is an unique identifier.
+
+We follow exactly the same steps to verify that a document has not been tampered.
+
+The [document store](/docs/verifiable-document/document-store) is a smart contract on the Ethereum network that records the issuance and revocation status of OpenAttestation documents. It will store hashes of wrapped documents that proves that the owner of the document store really issued the document.
+
+Imagine that you wrap a thousand of file and had to issue the `targetHash` for each of them. It would be extremely inefficient. That's where the `merkleRoot` will come handy.
+
+#### merkleRoot
+
+Once the `targetHash` of a document is computed, OpenAttestation will determine the `merkleRoot`. The `merkleRoot` value is the merkle root hash computed from the [merkle tree](https://en.wikipedia.org/wiki/Merkle_tree) using the `targetHash` of all the document wrapped together. After computing the merkle tree, the `merkleRoot` associated to a document will be added to it as well as the proofs (intermediate hashes) needed to ensure that the `targetHash` has been used to compute the `merkleRoot`. The proofs are added into the `proof` property.
```

Comment:
dunno if you wanna talk about what a merkle tree is?
and that each targetHash is a leaf in the tree

and computing the merkle tree using a whole batch of documents (targetHashes)

Reply:
can add that the targetHash is a leaf of a tree, about merkle tree I would prefer to link to a doc (if there is abything better than wiki)

Reply:
added `that each targetHash is a leaf in the tree`